File: pr-app/scripts/apply_auto_fixes.mjs
import fs from 'fs';
import * as parser from '@babel/parser';
import _traverse from '@babel/traverse';
import _generate from '@babel/generator';

const traverse = _traverse.default || _traverse;
const generate = _generate.default || _generate;

const autoFixFile = 'scan_report.json';
const report = JSON.parse(fs.readFileSync(autoFixFile, 'utf8'));

// Filter to only the files we want to auto-fix (mostly non_compliant_action_button)
const autoFixableFindings = report.findings.filter(item => {
    // Only process if all issues are non_compliant_action_button and it's not a warehouse nav tile
    const hasOnlyActionBtnIssues = item.issues.every(i => i.type === 'non_compliant_action_button');
    if (!hasOnlyActionBtnIssues) return false;
    
    // Ignore Warehouse navigation tiles
    if (item.file.includes('Warehouse') && item.issues.some(i => i.detail.includes('Manage') || i.detail.includes('Stock by') || i.detail.includes('HU to HU'))) {
        return false;
    }
    return true;
});

console.log(`Found ${autoFixableFindings.length} files eligible for auto-fix.`);

autoFixableFindings.forEach(finding => {
    const filePath = finding.file;
    const code = fs.readFileSync(filePath, 'utf8');

    const ast = parser.parse(code, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript']
    });

    let modified = false;

    traverse(ast, {
        JSXElement(path) {
            const el = path.node.openingElement;
            if (el.name.type === 'JSXIdentifier' && (el.name.name === 'button' || el.name.name === 'Button')) {
                // Determine if this is one of our target buttons.
                // It's hard to match exactly by text using AST easily, so we heuristically look for buttons with onClick
                // that trigger standard business logic, or we just rely on updating all primary action buttons in the form area.
                
                // For safety vs complexity of matching exact text, let's just add the missing classes to the specific button
                // if it resembles a primary action button (e.g., has 'onClick', not a small icon button)
                
                const hasOnClick = el.attributes.some(attr => attr.type === 'JSXAttribute' && attr.name.name === 'onClick');
                const isSubmit = el.attributes.some(attr => attr.type === 'JSXAttribute' && attr.name.name === 'type' && attr.value?.value === 'submit');
                
                // Quick hack: we'll look at the text content of the button to see if it matches the report detail
                const btnTextContent = generate(path.node.children.find(c => c.type === 'JSXText') || path.node).code;
                
                const isTarget = finding.issues.some(issue => {
                    const issueContent = issue.detail.split("'")[1] || "";
                    if (!issueContent) return false;
                    const normalizedIssue = issueContent.trim().replace(/\s+/g, ' ');
                    const normalizedBtn = btnTextContent.trim().replace(/\s+/g, ' ');
                    return normalizedBtn.includes(normalizedIssue) || normalizedIssue.includes('Confirm') || normalizedIssue.includes('Post') || normalizedIssue.includes('Add Item') || normalizedIssue.includes('Search');
                });

                if ((hasOnClick || isSubmit) && isTarget) {
                    let classNameAttr = el.attributes.find(attr => attr.type === 'JSXAttribute' && attr.name.name === 'className');
                    
                    if (classNameAttr && classNameAttr.value.type === 'StringLiteral') {
                        let classValue = classNameAttr.value.value;
                        if (!classValue.includes('w-full')) classValue += ' w-full';
                        if (!classValue.includes('bg-brand-blue')) classValue = classValue.replace(/bg-\w+-\d+/, '') + ' bg-brand-blue';
                        if (!classValue.includes('text-white')) classValue += ' text-white';
                        
                        classNameAttr.value.value = classValue.trim();
                        modified = true;
                    } else if (classNameAttr && classNameAttr.value.type === 'JSXExpressionContainer') {
                        // Complex expressions like template literals. 
                        // Instead of parsing, we can just append to the end.
                        // Actually, babel generator modifies the AST safely.
                    }
                }
            }
        }
    });

    if (modified) {
        // Warning: Babel generator might lose original formatting. 
        // For auto-classes, using a regex search/replace in Python might actually be safer for formatting.
    }
});

File: pr-app/scripts/apply_auto_fixes.py
import json
import re
import os

with open('scan_report.json', 'r', encoding='utf-8') as f:
    report = json.load(f)

for item in report['findings']:
    file_path = item['file']
    
    # Only process if ALL issues are non_compliant_action_button
    issues = item['issues']
    has_only_action_btns = all(i['type'] == 'non_compliant_action_button' for i in issues)
    if not has_only_action_btns:
        continue
        
    # Ignore false positive navigation tiles in Warehouse screens
    if 'Warehouse' in file_path and any('Manage' in i['detail'] or 'Stock by' in i['detail'] or 'HU to HU' in i['detail'] for i in issues):
        continue

    if not os.path.exists(file_path):
        continue

    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    modified = False

    for issue in issues:
        # Extract line content from the snippet to help find it
        snippet = issue['snippet'].strip()
        # Find the line number from the snippet if possible (Line X:)
        match = re.search(r'Line\s+(\d+):', issue['snippet'])
        if match:
            line_idx = int(match.group(1)) - 1
            if 0 <= line_idx < len(lines):
                line = lines[line_idx]
                
                # Check if it has a className attribute
                if 'className="' in line:
                    # Identify missing classes
                    missing_classes = []
                    if 'w-full' in issue['detail'] and 'w-full' not in line:
                        missing_classes.append('w-full')
                    if 'bg-brand-blue' in issue['detail'] and 'bg-brand-blue' not in line:
                        missing_classes.append('bg-brand-blue')
                    if 'text-white' in issue['detail'] and 'text-white' not in line:
                        missing_classes.append('text-white')

                    if missing_classes:
                        # Add them inside the className string
                        # e.g. className="btn px-4 bg-blue-500" -> className="btn px-4 bg-brand-blue text-white w-full"
                        
                        # First, remove incorrect background colors
                        line = re.sub(r'bg-\w+-\d+', '', line)
                        # Remove incorrect text colors (like text-white/50 if text-white is needed, though maybe risky, let's just append)
                        
                        insert_str = ' ' + ' '.join(missing_classes)
                        
                        # Insert right after className="
                        line = line.replace('className="', f'className="{insert_str.strip()} ', 1)
                        # Clean up double spaces
                        line = re.sub(r'\s+', ' ', line)
                        # Restore indentation
                        indent = len(lines[line_idx]) - len(lines[line_idx].lstrip())
                        lines[line_idx] = (' ' * indent) + line.strip() + '\n'
                        modified = True
                elif 'className={`' in line:
                    # Template literal classes
                    missing_classes = []
                    if 'w-full' in issue['detail'] and 'w-full' not in line:
                        missing_classes.append('w-full')
                    if 'bg-brand-blue' in issue['detail'] and 'bg-brand-blue' not in line:
                        missing_classes.append('bg-brand-blue')
                    if 'text-white' in issue['detail'] and 'text-white' not in line:
                        missing_classes.append('text-white')

                    if missing_classes:
                        # Append before the closing backtick
                        insert_str = ' ' + ' '.join(missing_classes)
                        
                        # First, remove incorrect background/text colors in the static part
                        line = re.sub(r'bg-(slate|gray|blue)-\d+', '', line)
                        
                        line = line.replace('`} ', f'{insert_str}`}} ')
                        line = line.replace('`}>', f'{insert_str}`}}>')
                        
                        indent = len(lines[line_idx]) - len(lines[line_idx].lstrip())
                        lines[line_idx] = (' ' * indent) + line.strip() + '\n'
                        modified = True

    if modified:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.writelines(lines)
        print(f"Auto-fixed {file_path}")

print("Auto-fix script completed.")
